File: src/main/java/com/neo/caption/ocr/service/impl/LoaderServiceImpl.java
package com.neo.caption.ocr.service.impl;

import com.neo.caption.ocr.property.AppInfoProperties;
import com.neo.caption.ocr.service.LoaderService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.bytedeco.javacpp.Loader;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.io.IOException;

@Service
@Slf4j
@RequiredArgsConstructor
public class LoaderServiceImpl implements LoaderService {

    private final AppInfoProperties appInfoProperties;

    @Override
    @Async
    public void loadLib(Class<?>... classes) {
        //noinspection ConfusingArgumentToVarargsMethod
        log.info("loading: {}", classes);
        Loader.load(classes, false);
    }

    @Override
    @Async
    public void logInfo() {
        log.info("system info: [platform={}, chips={}, core={}, processors={}]",
                Loader.getPlatform(), Loader.totalChips(), Loader.totalCores(), Loader.totalProcessors());
        try {
            log.info("javacpp info: [cache={}, temp={}, version={}]",
                    Loader.getCacheDir(), Loader.getTempDir(), Loader.getVersion());
        } catch (IOException e) {
            log.error("Failed to get javacpp information");
            e.printStackTrace();
        }
        log.info("app info: [JavaVersion={}, SpringBootVersion={}, appVersion={}, build={}]",
                appInfoProperties.getJavaVersion(), appInfoProperties.getSpringBootVersion(),
                appInfoProperties.getVersion(), appInfoProperties.getBuildTimestamp());
    }

}

File: src/main/kotlin/com/neo/caption/ocr/service/impl/LoaderServiceImpl.kt
package com.neo.caption.ocr.service.impl

import com.neo.caption.ocr.annotation.Slf4j
import com.neo.caption.ocr.annotation.Slf4j.Companion.log
import com.neo.caption.ocr.property.AppInfoProperties
import com.neo.caption.ocr.service.LoaderService
import org.bytedeco.javacpp.Loader
import org.springframework.core.env.Environment
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import java.net.InetAddress
import kotlin.system.exitProcess

@Slf4j
@Service
class LoaderServiceImpl(
    private val appInfoProperties: AppInfoProperties,
    private val environment: Environment,
) : LoaderService {

    @Async
    override fun loadLib(vararg classes: Class<*>) {
        classes.forEach {
            Loader.load(it)
            log.info { "loading class: $it" }
        }
    }

    @Suppress("HttpUrlsUsage")
    @Async
    override fun printSystemInfo() {
        try {
            log.info {
                "system info: [" +
                        "platform=${Loader.getPlatform()}, " +
                        "chips=${Loader.totalChips()}, " +
                        "core=${Loader.totalCores()}, " +
                        "processors=${Loader.totalProcessors()}" +
                        "]"
            }
            log.info {
                "javacpp info: [" +
                        "cacheDir=${Loader.getCacheDir()}, " +
                        "tempDir=${Loader.getTempDir()}, " +
                        "version=${Loader.getVersion()}" +
                        "]"
            }
        } catch (e: Throwable) {
            e.printStackTrace()
            log.error("Failed to get javacpp information, server will exit!!")
            exitProcess(-1)
        }
        log.info {
            "app info: [" +
                    "JavaVersion=${appInfoProperties.javaVersion}, " +
                    "SpringBootVersion=${appInfoProperties.springBootVersion}, " +
                    "appVersion=${appInfoProperties.version}, " +
                    "build=${appInfoProperties.buildTimestamp}" +
                    "]"
        }
        val ip = InetAddress.getLocalHost().hostAddress
        val port = environment.getProperty("server.port")
        log.info { "COCR server is ready! Open http://$ip:$port in your browser" }
    }

}